File: ZLForm/Classes/ZLFormRowDescriptor.m
//
//  ZLFormRowDescriptor.m
//  ZLForm
//
//  Created by admin on 2026/5/15.
//

#import "ZLFormRowDescriptor.h"
#import "ZLFormSectionDescriptor.h"
#import "ZLFormBaseCell.h"
#import "ZLFormValidator.h"

@interface ZLFormRowDescriptor ()
@property (nonatomic, strong,readwrite) ZLFormBaseCell *cell;
@property (nonatomic, strong) NSMutableArray<ZLFormValidator *> *validators;
@end
@implementation ZLFormRowDescriptor
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.configureCellBlock = ^(UITableViewCell * _Nonnull cell, id  _Nonnull value, ZLFormRowDescriptor * _Nonnull rowDescriptor) {
            
        };
        self.updateCellBlock = ^(UITableViewCell * _Nonnull cell, id  _Nonnull value, ZLFormRowDescriptor * _Nonnull rowDescriptor) {
            
        };
        [self addObserver:self
               forKeyPath:@"value"
                  options:NSKeyValueObservingOptionOld | NSKeyValueObservingOptionNew context:0];
    }
    return self;
}
- (NSMutableArray *)validators {
    if (!_validators) {
        _validators = [NSMutableArray array];
    }
    return _validators;
}
+(instancetype)formRowDescriptorWithTag:(NSString *)tag {
    return [[self alloc] initWithTag:tag];
}
- (ZLFormBaseCell *)cell {
    if (!_cell) {
        _cell = [self createCell];
        _cell.rowDescriptor = self;
    }
    return _cell;
}
- (ZLFormBaseCell *)createCell {
    if (self.cellClass) {
        return [[self.cellClass alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:self.tag];
    }else {
        return [[ZLFormBaseCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:self.tag];
    }
}
- (CGFloat)height {
    CGFloat height = _height;
    if (self.cell) {
        if ([self.cell conformsToProtocol:@protocol(ZLFormDescriptorCell)]) {
            if ([self.cell respondsToSelector:@selector(cellHeightForRowDescriptor:)]) {
                height = [(id<ZLFormDescriptorCell>)self.cell cellHeightForRowDescriptor:self];
            }
        }
    }
    return height > 0 ? height : 44.0;
}

-(instancetype)initWithTag:(NSString *)tag  {
    self = [self init];
    if (self) {
        self.tag = tag;
    }
    return self;
}
-(UITableViewCell<ZLFormDescriptorCell> *)cellForFormController:(UIViewController *)formController {
    return self.cell;
}

- (id )valueForDisplay {
    if (self.valueMapperToDisplay) {
        return  self.valueMapperToDisplay(self.value);
    }
    return self.value;
}
- (id )valueForStorage {
    if (self.storageValueMapper) {
        return self.storageValueMapper(self.value);
    }
    return self.value;
}
#pragma mark - KVO

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
    if (!self.sectionDescriptor) return;
    if (object == self && [keyPath isEqualToString:@"value"]) {
        id newValue = [change objectForKey:NSKeyValueChangeNewKey];
        id oldValue = [change objectForKey:NSKeyValueChangeOldKey];
        if ([keyPath isEqualToString:@"value"]) {
            if (self.onChangeBlock) {
                self.onChangeBlock(oldValue, newValue, self);
            }
        }
    }
}
- (void)dealloc
{
    [self removeObserver:self forKeyPath:@"value"];
}

- (void)addValidator:(ZLFormValidator *)validator{
    if (validator) {
        [self.validators addObject:validator];
    }
}
- (void)removeValidator:(ZLFormValidator *)validator {
    if (validator) {
        [self.validators removeObject:validator];
    }
}
- (void)addValidator:(NSString *)msg validationBlock:(BOOL (^)(id value))validationBlock {
    ZLFormValidator *validator = [[ZLFormValidator alloc] initWithMsg:msg validationBlock:validationBlock];
    [self addValidator:validator];
}
- (ZLFormValidationStatus *)doValidation {
    for (ZLFormValidator *validator in self.validators) {
        ZLFormValidationStatus *status = [validator validate:self];
        if (!status.isValid) {
            return status;
        }
    }
    return [ZLFormValidationStatus formValidationStatusWithMsg:@"" status:YES rowDescriptor:self];
}
@end

File: ZLForm/Classes/ZLFormRowDescriptor.swift
import UIKit
import DifferenceKit
public typealias ZLOnChangeBlock = (_ oldValue: Any?, _ newValue: Any?, _ rowDescriptor: ZLFormRowDescriptor) -> Void
public typealias ZLConfigureCellBlock = (_ cell: UITableViewCell, _ value: Any?, _ rowDescriptor: ZLFormRowDescriptor) -> Void
public typealias ZLUpdateCellBlock = (_ cell: UITableViewCell, _ value: Any?, _ rowDescriptor: ZLFormRowDescriptor) -> Void
public typealias ZLCellProviderBlock = (_ rowDescriptor: ZLFormRowDescriptor) -> UITableViewCell

public typealias ZLValueMapperBlock = (_ row: ZLFormRowDescriptor,_ value: Any?) -> Any?

///row被点击
public typealias ZLFormRowDidSelectBlock = (_ rowDescriptor: ZLFormRowDescriptor, _ indexPath: IndexPath) -> Void


@objcMembers
public class ZLFormRowDescriptor: NSObject, Differentiable {
    
    public var cellClass: AnyClass?
    public var cellProvider: ZLCellProviderBlock?
    public var height: CGFloat = 0
    public var tag: String = ""
    public var key: String?
    public var title: String?
    @objc public dynamic var value: Any?
    public weak var sectionDescriptor: ZLFormSectionDescriptor?
    
    public var onChangeBlock: ZLOnChangeBlock?
    public var configureCellBlock: ZLConfigureCellBlock?
    public var updateCellBlock: ZLUpdateCellBlock?
    public var didSelectBlock: ZLFormRowDidSelectBlock?
    
    public var disabled: Bool = false
    public var ignoreValue: Bool = false

    
    public var valueMapperToDisplay: ZLValueMapperBlock?
    public var storageValueMapper: ZLValueMapperBlock?
    public var placeholderValue: Any?
    public var hidden: Bool = false

    /// cell 展示出来的时候indexPath 会被赋值，默认为 nil，ZLFormDescriptor 会在展示 cell 的时候赋值为当前 cell 在 tableView 中的 indexPath。
    public internal(set) var indexPath: IndexPath?
    
    private var _cell: UITableViewCell?
    private var validators: [ZLFormValidator] = []
    private var observation: NSKeyValueObservation?
    
    public var cell: UITableViewCell  {
        if _cell == nil {
            let cell = createCell()
            if let cell = cell as? ZLFormDescriptorCell {
                cell.rowDescriptor = self
            }
           _cell = cell
        }
        return _cell!
    }
    // MARK: - Init
    
    private override init() {
        super.init()
        setupKVO()
    }
    
    public init(tag: String) {
        super.init()
        self.tag = tag
        setupKVO()
    }
    
    @objc public class func formRowDescriptor(tag: String) -> ZLFormRowDescriptor {
        return ZLFormRowDescriptor(tag: tag)
    }
    private func setupKVO() {
        observation = observe(\.value, options: [.old, .new]) { [weak self] _, change in
            guard let self = self, self.sectionDescriptor != nil else { return }
            let oldVal = change.oldValue as? Any
            let newVal = change.newValue as? Any
            self.onChangeBlock?(oldVal, newVal, self)
        }
    }
    
    // MARK: - Cell
    
    private func createCell() -> UITableViewCell  {
        if let provider = cellProvider  {
            return provider(self)
        }
        if let cls = cellClass as? UITableViewCell.Type {
            let cell = cls.init(style: .default, reuseIdentifier: tag)
            return cell
        }
        return ZLFormBaseCell(style: .default, reuseIdentifier: tag)
    }
    
    @objc public func cellForFormController(_ formController: UIViewController) -> UITableViewCell  {
        return cell
    }
    public func effectiveHeight() -> CGFloat {
        if let formCell = cell as? ZLFormDescriptorCell, let cellHeight = formCell.cellHeight?(for: self), cellHeight > 0 {
            return cellHeight
        }
        
        if height > 0 {
            return height
        }
        return UITableView.automaticDimension
    }
    
    // MARK: - Value
    
    @objc public func valueForDisplay() -> Any? {
        if let mapper = valueMapperToDisplay {
            return mapper(self,value)
        }
        return value
    }
    
    @objc public func valueForStorage() -> Any? {
        if let mapper = storageValueMapper {
            return mapper(self,value)
        }
        return value
    }
    
    // MARK: - Validators
    
    public func addValidator(_ validator: ZLFormValidator) {
        validators.append(validator)
    }
    
    public func removeValidator(_ validator: ZLFormValidator) {
        validators.removeAll { $0 === validator }
    }
    
    @objc public func addValidator(_ msg: String, validationBlock: @escaping (Any?) -> Bool) {
        let validator = ZLFormValidator(msg: msg, validationBlock: validationBlock)
        addValidator(validator)
    }
    
    @objc public func doValidation() -> ZLFormValidationStatus {
        for validator in validators {
            let status = validator.validate(self)
            if !status.isValid {
                return status
            }
        }
        return ZLFormValidationStatus(msg: "", status: true, rowDescriptor: self)
    }
    
    
    // MARK: - Differentiable
    private let uuid = UUID().uuidString
    
    public var differenceIdentifier: String {
        return tag.isEmpty ? uuid : tag
    }
    
    public var contentEqualHandler: ((ZLFormRowDescriptor, ZLFormRowDescriptor) -> Bool)?

    public func isContentEqual(to source: ZLFormRowDescriptor) -> Bool {

        if let handler = contentEqualHandler ?? source.contentEqualHandler {
            return handler(self, source)
        }
        return tag == source.tag &&
               title == source.title &&
               (value as? NSObject)?.isEqual(source.value) == true
    }
}
